geth/jail/internal/loop/loop.go

```diff
@@ -131,28 +176,30 @@ func (l *Loop) processTask(t Task) error {
 // Run handles the task scheduling and finalisation.
 // It runs infinitely waiting for new tasks.
 func (l *Loop) Run(ctx context.Context) error {
+	defer l.close()
+	defer l.removeAll()
+
 	for {
 		select {
 		case t := <-l.ready:
 			if ctx.Err() != nil {
-				l.removeAll()
 				return ctx.Err()
 			}
 
 			if t == nil {
 				continue
 			}
 
-			err := l.processTask(t)
-			if err != nil {
-				// TODO(divan): do we need to report
-				// errors up to the caller?
-				// Ignoring for now, as loop
-				// should keep running.
-				continue
-			}
+			go func() {
```

Comment:
Okay, this fundamentally changes the idea of `loop` as it refers to `event loop` — one single loop that guarantees that only one task is executed at a time (limitation of Otto JS VM, basically). So we have to run this synchronously.
I saw your comment about possible synchronization issue here between `Ready()` and `processTask()`, but I'm not sure I got it. Can you elaborate? In `Ready()` mutex should be released before writing into channel. Also, we can use `NewWithBacklog()` to use buffered channel in `Cell`.

Reply:
Right now the mutex is not released before writing into the channel, and that is the issue.

The default is to create an unbuffered channel. But if we do that any a few tasks are readied in a row, then we will begin to process the the first task as more tasks are attempting to be added to the channel. Once a task is processed, `(*loop).removeByID` is called, which tries to acquire the lock. However, it cannot because `(*loop).Ready` has it and won't release it until it is able to put another task on the channel. We achieve deadlock.

For example, without having `(*loop).processTask` in its own goroutine, tests like `TestFetchWithHandlerParallel` in `fetch_test.go` will fail.

A possible solution would be to have 
```
func (l *Loop) Ready(t Task) error {
    l.lock.RLock()
    if !l.accepting {
        t.Cancel()
        l.lock.RUnlock()
	return ErrClosed
    }
    l.lock.RUnlock()
    l.ready <- t
    return nil
}
```
Not elegant code, but it would work.

Reply:
Perhaps a more elegant solution:
```
func (l *Loop) Ready(t Task) error {
	var err error
	l.lock.RLock()
	defer func() {
		l.lock.RUnlock()
		if err == nil {
			l.ready <- t
		}
	}()
	if !l.accepting {
		t.Cancel()
		err = ErrClosed
	}
	return err
}
```

Reply:
Yeah, that's more or less how it might work.

But it's imporant that `processTask` should be run syncronously in the loop as this whole package implements JS VM event loop and otto VM can't execute tasks concurrently (it may actually panic if we run similar code on the same VM instance concurrently).